### PR TITLE
Implemented caching for Party model

### PR DESCRIPTION
**Pull request by angushtlam**
title uwu

* Implemented caching
* Also fixed the CachedModel abstract class to support typing

### CODE REVIEW

File: src/main/java/com/raeic/embarker/party/commands/JoinCommand.java [hunk 1]
```diff
@@ -108,6 +109,9 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
                 PartyPlayer newPartyPlayer = new PartyPlayer(playerUniqueId, inviteSenderPlayerUniqueId);
                 newPartyPlayer.save();
 
+                // Invalidate the existing party as we updated it.
+                Globals.party.invalidateCacheByKey(inviteSenderPlayerUniqueId);
```
Comment (nenofite) [inline src/main/java/com/raeic/embarker/party/commands/JoinCommand.java]: What if we moved this into PartyPlayer#save? Theoretically whenever a PartyPlayer is saved, the cache will be invalidated, right?

**Review by nenofite — CHANGES_REQUESTED** [review]

File: src/main/java/com/raeic/embarker/party/commands/JoinCommand.java (hunk 1 above)
Comment (angushtlam) [reply src/main/java/com/raeic/embarker/party/commands/JoinCommand.java]: thats true is a good idea

File: src/main/java/com/raeic/embarker/party/models/PartyPlayer.java [hunk 2]
```diff
@@ -20,6 +20,10 @@ public String getLeaderUniqueId() {
     }
 
     public void delete() {
+        // Invalidate the cache on the party leader's Party object, as the
+        // party's been updated.
+        Globals.party.invalidateCacheByKey(leaderUniqueId);
```
Comment (nenofite) [inline src/main/java/com/raeic/embarker/party/models/PartyPlayer.java]: There's a race condition here, we need to invalidate the cache after it's deleted from SQL

File: src/main/java/com/raeic/embarker/party/models/PartyPlayer.java [hunk 3]
```diff
@@ -37,6 +41,10 @@ public void delete() {
     }
 
     public void save() {
+        // Invalidate the cache on the party leader's Party object, as the
+        // party's been updated.
+        Globals.party.invalidateCacheByKey(leaderUniqueId);
```
Comment (nenofite) [inline src/main/java/com/raeic/embarker/party/models/PartyPlayer.java]: ditto

**Review by nenofite — CHANGES_REQUESTED** [review]

**Review by nenofite — APPROVED** [review]
ty for fixing 🚀

**Pull request merged**